### PR TITLE
Role Config documentation

### PR DESCRIPTION
**Pull request by 5abeel**

### CODE REVIEW

File: docs/guides/p4-role-configuration.rst [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+   exclusive_p4_ids: 48773578  // table ID for MainControlDecrypt.lem_exception
+   exclusive_p4_ids: 38116127  // table ID for MainControlDecrypt.lem_clear
+   exclusive_p4_ids: 45068641  // table ID for MainControlDecrypt.ipsec_rx_sa_classification_table
+   exclusive_p4_ids: 41233864  // table ID for crypto_tunnel_control.ipsec_rx_post_decrypt_table
+   ...
+   receives_packet_ins: false
+   can_push_pipeline: false
+
+
+4. Connect P4 client to infrap4d
```
Comment (ffoulkes) [inline docs/guides/p4-role-configuration.rst:88]: ditto

File: docs/guides/p4-role-configuration.rst [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+Once the configuration is complete, the P4 clients will push a role name and
+the role configuration file to infrap4d with a `MasterArbitrationUpdate` message
```
Comment (ffoulkes) [inline docs/guides/p4-role-configuration.rst:90]:
```text
the role configuration file to infrap4d with a ``MasterArbitrationUpdate`` message
```

File: docs/guides/p4-role-configuration.rst [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+after the gRPC connection is initiated.
+
+Your Own P4 Client
```
Comment (ffoulkes) [inline docs/guides/p4-role-configuration.rst:93]:
```suggestion
Your P4 Client
```

File: docs/guides/p4-role-configuration.rst [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+~~~~~~~~~~~~~~~~~~
+
+If using your own P4 client, you can set role name and pack `Any` protobuf
+message with the role configuration. Details of how to perform this are
+available on this page:
```
Comment (ffoulkes) [inline docs/guides/p4-role-configuration.rst:98]:
```suggestion
available in this document:
```
This is a reference to an external document, not a page in the user guide.

I would reserve _page_ for the latter

File: docs/guides/p4-role-configuration.rst [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+https://github.com/ipdk-io/stratum-dev/blob/split-arch/stratum/public/proto/p4_role_config.md
```
Comment (ffoulkes) [inline docs/guides/p4-role-configuration.rst:99]:
```suggestion
https://github.com/ipdk-io/stratum-dev/blob/split-arch/stratum/public/proto/p4_role_config.md
```
```suggestion
https://github.com/ipdk-io/stratum-dev/blob/split-arch/stratum/public/proto/p4_role_config.md.
```
End of sentence.

**Review by ffoulkes — COMMENTED** [review]
Right now, this is an orphan document. You need to add it to an index file.

I would suggest putting it in `docs/index.rst`. Create a **Guides** section between the **Clients** and **P4 Programs** sections, and make the new document part of it.

```text
.. toctree::
   :maxdepth: 1
   :caption: Guides

   guides/p4-role-configuration.rst
```
This isn't a perfect solution, but it's an okay start. The structure of the user guide evolves as its contents take shape.

File: docs/guides/p4-role-configuration.rst [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+1. Compile P4 program
+Compile the application program according to the instructions in Compiling P4
+programs guide to generate P4 artifacts for programming the pipeline.
+
+2. Extract table IDs
+From the P4 artifact `p4info.txt`, extract P4 table IDs. A helper script is
+provided to assist in extracting all table IDs from the provided input file.
+
+Each P4 client will need its own role configuration, so run the script twice to
```
Comment (ffoulkes) [inline docs/guides/p4-role-configuration.rst:43]:
```suggestion
Each P4 client will need its own role configuration, so run the script once for each of them.
```

File: docs/guides/p4-role-configuration.rst [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+4. Connect P4 client to infrap4d
+Once the configuration is complete, the P4 clients will push a role name and
+the role configuration file to infrap4d with a `MasterArbitrationUpdate` message
+after the gRPC connection is initiated.
+
+Your Own P4 Client
+~~~~~~~~~~~~~~~~~~
+
+If using your own P4 client, you can set role name and pack `Any` protobuf
+message with the role configuration. Details of how to perform this are
```
Comment (ffoulkes) [inline docs/guides/p4-role-configuration.rst:97]: `Any` is a field type, isn't it? And don't messages and fields have names? And if one "packs" a field, what's the type of the thing you're storing in the Any field.

This sentence feels somewhat vague.

File: docs/guides/p4-role-configuration.rst [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+Example Configuration
+~~~~~~~~~~~~~~~~~~~~~
+
+This example demonstrates two control planes (i.e., two P4 clients)
+partitioning the pipeline into non-overlapping sections giving each client
+exclusive access to specified P4 tables. Clients chosen for this example are
+Linux Networking and IPsec.
+
+1. Compile P4 program
+Compile the application program according to the instructions in Compiling P4
```
[the remark this answers is not in the document]
Comment (5abeel) [reply docs/guides/p4-role-configuration.rst:36]: Yeah, I had a problem with adding hyperlinks in rst document. The linters fail when I tried adding a link saying that I was adding a .md format style links. TBD to figure out how to insert links in future

File: docs/guides/p4-role-configuration.rst (hunk 7 above)
Comment (5abeel) [reply docs/guides/p4-role-configuration.rst:97]: I didnt want to re-write or copy-paste the Stratum text here, so just left a quick note and a link to the document where Stratum documentation page shows how to pack the `Any` field. Hope this is sufficient for this document.

**Comment by 5abeel** [comment]
All comments addressed. Please review

**Review by ffoulkes — APPROVED** [review]
LGTM (but...)

All four of your numbered paragraphs are number 1. These are paragraphs, not headings, so rst won't magically renumber them for you. You have to do this manually.

You might want to fix this before you merge.

image

**Comment by 5abeel** [comment]
The rst check fails with an ordered number specified. I ended up turning it into an unordered list instead

**Comment by ffoulkes** [comment]
> The rst check fails with an ordered number specified. I ended up turning it into an unordered list instead

Oh. I know what the problem is.

Each line beginning with a number is seen as the start of an ordered list. The following paragraph is indented, so it concludes that the list is done. The next line beginning with a number is the start of a _new_ list. And new lists must start with item number 1.

**Review by nupuruttarwar — APPROVED** [review]
LGTM

**Pull request merged**